src/storage/compression/numeric_constant.cpp

```diff
@@ -0,0 +1,158 @@
+#include "duckdb/storage/statistics/validity_statistics.hpp"
+#include "duckdb/storage/table/column_segment.hpp"
+#include "duckdb/function/compression_function.hpp"
+#include "duckdb/storage/segment/uncompressed.hpp"
+
+namespace duckdb {
+
+//===--------------------------------------------------------------------===//
+// Scan
+//===--------------------------------------------------------------------===//
+unique_ptr<SegmentScanState> ConstantInitScan(ColumnSegment &segment) {
+	return nullptr;
+}
+
+//===--------------------------------------------------------------------===//
+// Scan base data
+//===--------------------------------------------------------------------===//
+void ConstantScanFunctionValidity(ColumnSegment &segment, ColumnScanState &state, idx_t scan_count, Vector &result) {
+	auto &validity = (ValidityStatistics &)*segment.stats.statistics;
+	if (validity.has_null) {
+		result.SetVectorType(VectorType::CONSTANT_VECTOR);
+		ConstantVector::SetNull(result, true);
+	}
+}
+
+template <class T>
+void ConstantScanFunction(ColumnSegment &segment, ColumnScanState &state, idx_t scan_count, Vector &result) {
+	auto &nstats = (NumericStatistics &)*segment.stats.statistics;
+
+	auto data = FlatVector::GetData<T>(result);
+	data[0] = nstats.min.GetValueUnsafe<T>();
+	result.SetVectorType(VectorType::CONSTANT_VECTOR);
+}
+
+//===--------------------------------------------------------------------===//
+// Scan Partial
+//===--------------------------------------------------------------------===//
+void FillFunctionValidity(ColumnSegment &segment, Vector &result, idx_t start_idx, idx_t count) {
+	auto &validity = (ValidityStatistics &)*segment.stats.statistics;
+	if (validity.has_null) {
+		auto &mask = FlatVector::Validity(result);
+		for (idx_t i = 0; i < count; i++) {
+			mask.SetInvalid(start_idx + i);
+		}
+	}
+}
+
+template <class T>
+void FillFunction(ColumnSegment &segment, Vector &result, idx_t start_idx, idx_t count) {
+	auto &nstats = (NumericStatistics &)*segment.stats.statistics;
+
+	auto data = FlatVector::GetData<T>(result);
+	auto constant_value = nstats.min.GetValueUnsafe<T>();
+	for (idx_t i = 0; i < count; i++) {
+		data[start_idx + i] = constant_value;
+	}
+}
+
+void ConstantScanPartialValidity(ColumnSegment &segment, ColumnScanState &state, idx_t scan_count, Vector &result,
+                                 idx_t result_offset) {
+	FillFunctionValidity(segment, result, result_offset, scan_count);
+}
+
+template <class T>
+void ConstantScanPartial(ColumnSegment &segment, ColumnScanState &state, idx_t scan_count, Vector &result,
+                         idx_t result_offset) {
+	FillFunction<T>(segment, result, result_offset, scan_count);
+}
+
+//===--------------------------------------------------------------------===//
+// Fetch
+//===--------------------------------------------------------------------===//
+void ConstantFetchRowValidity(ColumnSegment &segment, ColumnFetchState &state, row_t row_id, Vector &result,
```

Comment:
Since a lot of the code checks/optimises for constant columns, does it make sense to scan into a constant column?

Reply:
The regular scan does exactly that (`ConstantScanFunction`). The fetch fetches only a single row, and it might be combined with calls to Fetch to other segments, so it cannot return a constant vector and instead needs to read only a single value into the row_id.